### PR TITLE
Add Credits File Generator

### PR DESCRIPTION
**Pull request by spen**
This PR seeks to introduce a utility script that generates a formatted `CREDITS.md` file containing each dependency found in `package.json` along with a link and license label for each.

### Motivations

I originally became interested in the `CREDITS.md` file while working on #12430, specifically while looking at merging `wp-desktop`s credits in to Calypsos `CREDITS.md` - The two files were formatted quite differently to each other.
Personally, I find the current format of Calypsos `CREDITS.md` too verbose to be useful - it's not possible to skim the file to see which projects we are giving credit to because each one is broken up with its license.
Another issue that some folk have mentioned is that this file also needs to be maintained manually - each time a new dependency is added, the author must remember to update this document. Often this step is forgotten, leaving the document incomplete.

### Delphin To The Rescue!

99.9% of the credit for this PR goes to @gziolo & @aidvu as this work is near-enough a lift & shift from https://github.com/Automattic/delphin/pull/1153. 
In that PR a script was added that generates a new `CREDITS.md` file based on the dependencies found in the `package.json`, using [`nlf`](https://github.com/iandotkelly/nlf) to retrieve licenses for each package.
The formatting for Delphins `CREDITS.md` file is quite clean. It groups each package by licence and provides a link to each packages repository.
Following this formatting with this PR, the resulting document is reduced from 1400+ lines to ~200 and is much more easy to reference.

### It's Not Perfect 🙁
Unfortunately there are some kinks with certain packages:
* `tinymce` shows `( none )` instead of a repo link.
* `twemoji` shows up under the license type: `MIT, undefined`.
* A number of packages show the `ssh` URL and so are not linkable.

I'm thinking about making packages overridable - it would fix these caveats but is less than ideal.

### Testing
Testing should be as simple as running:
```make
make credits
```
from the project root and checking through the outputted `CREDITS.md`

### Discussion Points
* Do we actually want a new format?
* Is the format from Delphine the one we wish to settle with?
* Would a per-package override mechanism be a good addition?
* Should we go a little further and automate this with a git hook or allow it to be manual?

### CODE REVIEW

**Comment by gziolo** [comment]
> Would a per-package override mechanism be a good addition?

What I did with Delphin is I fixed repositories which had invalid license information exposed :) It's easier to open issue or PR in the repository than provide workarounds. `nlf` uses well established standards to find out license type.

**Comment by spen** [comment]
> What I did with Delphin is I fixed repositories which had invalid license information exposed

Good shout! I'll look at doing the same :)

**Comment by gziolo** [comment]
> Should we go a little further and automate this with a git hook or allow it to be manual?

`pre-commit hook` seems like an interesting approach, but it's fine to catch it at review, too. In the end we rarely modify `package.json` file. Another option would be have a bot to ping Calypso channel that it needs update.

**Comment by gziolo** [comment]
Do you still need to run `make shrinkwrap` when yo modify `devDependencies`? It would be nice to double check :)

**Comment by spen** [comment]
I did think that a hook may be overkill, especially since it does take some time to generate.
This does make me think that adding a step to `make credits` to first check for modifications to `package.json` would be helpful, though I'm not sure off the top of my head how we would reliably make that check.

**Comment by aidvu** [comment]
Nice work. `pre-commit` worth trying. :)

Other than that, I suggest contributing proper licensing to repos we use. For the `ssh` ones, hmm. I'd say it's a good enough reference, as all of these kind-of point to GH, and people looking at it will know where to find them.

**Comment by aidvu** [comment]
@spen any time to wrap this up? I can probably pick it up for trash pickup day, or we can work together to finish it?

**Comment by spen** [comment]
@aidvu I'd love to get this over line :) I've been a bit busy with support rotation lately though.
Do you have an idea of what's left to bring this to completion? 
Is it just a case of opening PR's in the erroneous repositories and creating a pre-commit hook (at least to try it out)?

**Comment by aidvu** [comment]
We should check with legal on it, and then submit PRs for the repos we use. I don't think anything should be blocking it. As for the auto-update, pre-commit might work (tho this takes a while to run, and I'm not sure if we want to do that). Probably should try a diff for changes first, then run the credits generation.

**Comment by spen** [comment]
I'm +1 for tackling separate :)

**Comment by aidvu** [comment]
Touch base next week. Will reserve time to get this in. :)

File: Makefile [hunk 1]
```diff
@@ -242,6 +242,9 @@ docker-run:
 	@echo running wp-calypso docker image at localhost:3000
 	docker run -it --name wp-calypso --rm -p 80:3000 -e NODE_ENV='wpcalypso' -e CALYPSO_ENV='wpcalypso' wp-calypso
 
+credits:
+	@$(NODE) bin/generate-credits.js
```
Comment (gziolo) [inline Makefile]: We can call `npm run credits` here.

**Comment by spen** [comment]
[Add Credits File Generator](https://trello.com/c/h5m6J5mG/278-add-credits-file-generator)

**Comment by alisterscott** [comment]
Closing as this PR is stale - please reopen if necessary

**Pull request closed without merging**